### PR TITLE
Add MemoizedCombinedRegexp

### PR DESCRIPTION
**Pull request by xepozz**

### CODE REVIEW

File: src/MemoizedCombinedRegexp.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+declare(strict_types=1);
+
+namespace Yiisoft\Strings;
+
+/**
+ * `CombinedRegexp` optimizes matching of multiple regular expressions.
+ * Read more about the concept in
+ * {@see https://nikic.github.io/2014/02/18/Fast-request-routing-using-regular-expressions.html}.
+ */
+final class MemoizedCombinedRegexp
```
Comment (vjik) [inline src/MemoizedCombinedRegexp.php:12]: Does it make sense to create an interface `CombinedRegexpInterface`?

File: src/MemoizedCombinedRegexp.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return $this->results[$string]['position'] ?? throw new \Exception(
+            sprintf(
+                'Failed to match pattern "%s" with string "%s".',
+                $this->getCompiledPattern(),
+                $string,
+            )
+        );
+    }
+
+    protected function evaluate(string $string): void
```
Comment (vjik) [inline src/MemoizedCombinedRegexp.php:75]:
```suggestion
    private function evaluate(string $string): void
```

**Review by vjik — CHANGES_REQUESTED** [review]
Overall looks good. But need:

- fix psalm errors
- add tests
- add info to readme
- add line to changelog

File: src/MemoizedCombinedRegexp.php (hunk 1 above)
Comment (xepozz) [reply src/MemoizedCombinedRegexp.php]: Created abstract class instead

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 ## 2.2.0 under development
 
+- New #103: Add `MemoizedCombinedRegexp` class (@xepozz)
```
Comment (vjik) [inline CHANGELOG.md:5]:
```suggestion
- New #103: Add `MemoizedCombinedRegexp` decorator that caches results of `CombinedRegexp` (@xepozz)
- Enh #103: Raise required PHP version to `^8.0` (@xepozz)
```

File: README.md [hunk 4]
```diff
@@ -21,11 +21,12 @@ The package provides:
 - `NumericHelper` that has static methods to work with numeric strings;
 - `Inflector` provides methods such as `toPlural()` or `toSlug()` that derive a new string based on the string given;
 - `WildcardPattern` is a shell wildcard pattern to match strings against.
```
Comment (vjik) [inline README.md:23]:
```suggestion
- `WildcardPattern` is a shell wildcard pattern to match strings against;
```

File: README.md [hunk 5]
```diff
@@ -21,11 +21,12 @@ The package provides:
...
-- `CombinedRegexp` is a wrapper that optimizes multiple regular expressions matching.
+- `CombinedRegexp` is a wrapper that optimizes multiple regular expressions matching
+  - `MemoizedCombinedRegexp` is a decorator that caches results of `CombinedRegexp` to speed up matching.
```
Comment (vjik) [inline README.md:25]:
```suggestion
- `CombinedRegexp` is a wrapper that optimizes multiple regular expressions matching and 
  `MemoizedCombinedRegexp` is a decorator that caches results of `CombinedRegexp` to speed up matching.
```

File: src/AbstractCombinedRegexp.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+     * Returns position of the pattern that matches the given string.
+     * @throws Exception if the string does not match any of the patterns.
+     */
+    abstract public function getMatchingPatternPosition(string $string): int;
+
+    /**
+     * @throws Exception
+     * @return never-return
+     */
+    protected function throwFailedMatchException(string $string): void
```
Comment (vjik) [inline src/AbstractCombinedRegexp.php:55]: This is once non-abstract method. May be convert class to interface?

File: src/MemoizedCombinedRegexp.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Strings;
+
+/**
+ * MemoizedCombinedRegexp is a decorator for {@see AbstractCombinedRegexp} that caches results of
```
Comment (vjik) [inline src/MemoizedCombinedRegexp.php:8]:
```suggestion
 * `MemoizedCombinedRegexp` is a decorator for {@see AbstractCombinedRegexp} that caches results of
```

File: src/Inflector.php [hunk 8]
```diff
@@ -296,7 +296,7 @@ final class Inflector
     /**
      * @param string[] $rules The rules for converting a word into its plural form.
      *
-     * @psalm-param array<string, string> $rules
+     * @psalm-param non-empty-array<non-empty-string, string> $rules
      * The keys are the regular expressions and the values are the corresponding replacements.
```
Comment (vjik) [inline src/Inflector.php:300]:
```suggestion
     * The keys are the regular expressions and the values are the corresponding replacements.
     * @psalm-param non-empty-array<non-empty-string, string> $rules
```

File: src/Inflector.php [hunk 9]
```diff
@@ -296,7 +296,7 @@ final class Inflector
     /**
      * @param string[] $rules The rules for converting a word into its plural form.
      *
-     * @psalm-param array<string, string> $rules
+     * @psalm-param non-empty-array<non-empty-string, string> $rules
```
Comment (vjik) [inline src/Inflector.php:299]: Should we check error on non-empty in this method?

File: tests/Support/TestCombinedRegexp.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Strings\Tests\Support;
+
+use Yiisoft\Strings\AbstractCombinedRegexp;
+
+class TestCombinedRegexp extends AbstractCombinedRegexp
```
Comment (vjik) [inline tests/Support/TestCombinedRegexp.php:9]: Seems, this class is never use.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/AbstractCombinedRegexp.php (hunk 6 above)
Comment (xepozz) [reply src/AbstractCombinedRegexp.php:55]: No need to keep it in the interface

File: src/Inflector.php (hunk 9 above)
Comment (xepozz) [reply src/Inflector.php]: I've just tried to fix psalm errors. If you know better way let's do it

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**